src/libPMacc/include/eventSystem/transactions/Transaction.tpp

```diff
@@ -60,15 +47,43 @@ inline EventTask Transaction::getTransactionEvent( )
     return baseEvent;
 }
 
-void Transaction::operation( ITask::TaskType )
+void Transaction::operation( ITask::TaskType operation )
 {
-    if ( isAtomic == false )
-        baseEvent.waitForFinished( );
+    if ( operation == ITask::TASK_CUDA )
+    {
+        Manager &manager = Environment<>::get( ).Manager( );
+
+        ITask* baseTask = manager.getITaskIfNotFinished( this->baseEvent.getTaskId( ) );
+        if ( baseTask != NULL )
+        {
+            if ( baseTask->getTaskType( ) == ITask::TASK_CUDA )
+            {
+                /* no blocking is needed */
+                return;
+            }
+        }
+    }
+    baseEvent.waitForFinished( );
 }
 
 EventStream* Transaction::getEventStream( ITask::TaskType )
 {
-    return this->eventStream;
+    Manager &manager = Environment<>::get( ).Manager( );
+    ITask* baseTask = manager.getITaskIfNotFinished( this->baseEvent.getTaskId( ) );
+
+    if ( baseTask != NULL )
+    {
+        if ( baseTask->getTaskType( ) == ITask::TASK_CUDA )
+        {
+            /* `StreamTask` from previous task must be reused to guarantee
```

Comment:
Actually the Stream from the StreamTask. What about those were you can have 2 streams? (The TaskAnd) there the stream is reset at some point and you'd probably need both streams, don't you?